src/protocol/abi/argument_encoding.md

```diff
@@ -0,0 +1,245 @@
+```
+
+## Fixed-length Strings
+
+Strings have fixed length and are encoded in-place. `str[n]`, where `n` is the fixed-size of the string. Rather than padding the string, the encoding of the elements of the string is tightly packed. And unlike the other type encodings, the string encoding is left-aligned.
+
+Note that all strings are encoded in UTF-8.
+
+**Example:**
+
+Encoding `"Hello, World"` as a `str[12]` **yields**:
+
+```text
+0x48656c6c6f2c20576f726c6400000000
+```
+
+Note that we're padding with zeroes in order to keep it right-aligned to 8 bytes (FuelVM's word size).
+
+## Structs
+
+Encoding ABIs that contain custom types, such as structs, is similar to encoding a set of primitive types. The encoding will proceed in the order that the inner types of a custom type are declared and _recursively_ just like encoding any other type. This way you can encode structs with primitive types or structs with more complex types in them such as other structs, arrays, strings, and enums.
+
+Here's an example:
+
+```rust
+struct InputStruct {
+    field_1: bool,
+    field_2: u8,
+}
+
+
+abi MyContract {
+    fn foo(a: u64);
+    fn bar(a: InputStruct);
+} {
+    fn baz(a: ()) { }
+}
+```
+
+Calling `bar` with `InputStruct{field_1: true, field_2: 5}` yields:
+
+```plaintext
+0x
+0000000000000001 // `true` encoded as a bool
+0000000000000005 // `5` encoded as u8
+```
+
+A more complex scenario:
+
+```rust
+struct InputStruct {
+    field_1: bool,
+    field_2: [u8; 2], // An array of u8, with length 2.
+}
+
+
+abi MyContract {
+    fn foo(a: u64);
+    fn bar(a: InputStruct);
+} {
+    fn baz(a: ()) { }
+}
+```
+
+Calling `bar` with `InputStruct{field_1: true, field_2: [1,2]}` yields:
+
+```plaintext
+0x
+0000000000000001 // `true` encoded as a bool
+0000000000000001 // `1` encoded as u8
+0000000000000002 // `2` encoded as u8
+```
+
+## Sum types (Enums)
+
+ABI calls containing sum types (enums) are encoded similarly to structs: encode the discriminant first, then recursively encode the type of the enum variant being passed to the function being called.
+
+```rust
+enum MySumType
+{
+    X: u32,
+    Y: bool,
+}
+
+abi MyContract {
+    fn foo(a: u64);
+    fn bar(a: MySumType);
+} {
+    fn baz(a: ()) { }
+}
+```
+
+Calling `bar` with `MySumType::X(42)` yields:
+
+```plaintext
+0x
+0000000000000000 // The discriminant of the chosen enum, in this case `0`.
+000000000000002a // `42` encoded as u64
+```
+
+If the sum type has variants of different sizes, then left padding is required.
+
+```rust
+enum MySumType
+{
+    X: b256,
+    Y: u32,
+}
+
+abi MyContract {
+    fn foo(a: u64);
+    fn bar(a: MySumType);
+} {
+    fn baz(a: ()) { }
+}
+```
+
+Calling `bar` with `MySumType::Y(42)` yields:
+
+```plaintext
+0x
+0000000000000001 // The discriminant of the chosen enum, in this case `1`.
+0000000000000000 // Left padding
+0000000000000000 // Left padding
+0000000000000000 // Left padding
+000000000000002a // `42` encoded as u64
```

Comment:
Does this correlate at all to how we encode them in Sway itself?  I'm pretty sure that the padding goes after the value in the memory layout in the compiler... but is that irrelevant to this ABI encoding stuff?

Also does the empty value for all-unit variants apply here or not?  I thought not, but... maybe?

Reply:
We have [this code](https://github.com/FuelLabs/sway/blob/4d5fea688f47617f9644deae387571e3140f5f50/sway-core/src/asm_generation/from_ir.rs#L254) in `from_ir.rs`. I remember aligning that with the spec. Same for all-unit variants and there are tests to make sure interfacing with the SDK is correct. We've made sure in the past that the memory layout is exactly the same internally and in the ABI.